Fixed bugs in handling race condition for verified status / Fixed a bug when domain mgr deletes container .aci file in Verified dir.. It should not.

kalyan-nidumolu opened the pull request:
1) Fixed a raceCondition in using VerifyImageStatus in zedmanager. Make sure there vs.RefCount > 0 OR there is a
    Verifier config already existing before using it. This is the main Fix. The rest are just cleanups.
        - Refactored processing of Images with VerifierStatus present into doInstallProcessStorageEntriesWithVerifiedImage
        - The real fix is really a few lines:
            if vs.RefCount == 0 && verifyImageConfig == nil:
                   return // Treat like verifier status not found. Re-download the image..
    
    2)     For containers, do not delete .aci file in the Verified directory from
    domainMgr..
    2) Fixed a logMfg from Errorf to Infof
    3) Added a couple more Info messages in Verifier to make it easy to debug.
    4) Removed uuidStr as an argument to functions taking AppInstanceStatus as another     argument. Passing uuidStr as another arg is superfluous in such cases - doUpdate, doInstall, doPrepare, doInactivateHalt
    5) Removed safename argument from MaybeAddVerifyImageConfig as it is part of StorageStatus already which is another arg.

eriknordmark commented on pkg/pillar/cmd/zedmanager/updatestatus.go:
```diff
@@ -371,10 +374,92 @@ func doUpdate(ctx *zedmanagerContext, uuidStr string,
+	}
+	if vs.RefCount == 0 {
+		log.Infof("Vs.RefCount = 0.")
+		// Check if Verifier Config exists.
```
Perhaps adding a useful comment referring to the cleanup handshake in handleverifier.go would be more useful for the person who needs to understand this code in then future.

eriknordmark commented on pkg/pillar/cmd/zedmanager/updatestatus.go:
```diff
@@ -462,55 +547,17 @@ func doInstall(ctx *zedmanagerContext, uuidStr string,
+			log.Debugf("doInstall: Verified image exists. StorageStatus "+
+				"Name: %s, safename %s, ImageSha256: %s",
+				ss.Name, safename, ss.ImageSha256)
+			changed = changed || statusUpdated
```
Shouldn't you update changed even if vs is nil? You shouldn't assume when doInstallProcessStorage... updates status, since that breaks modularity.

kalyan-nidumolu replied:
Ah.. I see what u r saying.. The intention is that statusUpdated can be true only when vs != nil.. 
Hmm.. What u say is actually better.. We are passing ssPtr..So we don't need such assumptions. That would keep it simple.. Fixed..

eriknordmark commented on pkg/pillar/cmd/zedmanager/updatestatus.go:
```diff
@@ -371,10 +374,92 @@ func doUpdate(ctx *zedmanagerContext, uuidStr string,
-	config types.AppInstanceConfig, status *types.AppInstanceStatus) (bool, bool) {
+// doInstallProcessStorageEntriesWithVerifiedImage
+//  Returns: (vs, storageStatusUpdate)
+func doInstallProcessStorageEntriesWithVerifiedImage(
```
Its a bit odd that this function can modify AppInstanceStatus; I was expecting it to only modify StorageStatus.
That should at least be documented in this comment.

Also, there is no point in returning a VerifyImageStatus since the caller doesn't use that. 

Hence the function definition is a bit wonky.
checkVerifyImageStatus() (bool, bool) i.e., return (found, changed)
makes more sense to me.

kalyan-nidumolu replied:
Thats a very observation. We actually don't see appInstanceStatus.ContainerId any more. Will remove the field. This also means I don;t need to pass AppInstanceStatus to the function. I just need the AppUUID.. Will fix it.

eriknordmark commented on pkg/pillar/cmd/zedmanager/updatestatus.go:
```diff
@@ -371,10 +374,92 @@ func doUpdate(ctx *zedmanagerContext, uuidStr string,
+		log.Infof("Vs.RefCount = 0.")
+		// Check if Verifier Config exists.
+		vcPtr := lookupVerifyImageConfig(ctx, safename)
+		if vcPtr == nil {
```
Given that handleverifier.go check status.Expired, why not check this here has well?
If Expired is set and there is no VerifyImageConfig, then we know that the verifier has been told it can delete the image.

kalyan-nidumolu replied:
Fixed.. Removed the vcPtr check.. and instead just checked foor Expired state as discussed. With Expired check.. we no longer need the config check.. If Expired. Just trigger re-download.

kalyan-nidumolu commented:
> You really want to make it painful to code review by including random cleanup so that the important code changes are harder to find ...
> 
> Why don't you need to remove the ACIDir? Where does it get removed?
> 

[Kalyan] As discussed, Verifier removes it..

> There are two calls to lookupVerifyImageStatusAny() but you are only fixing one of them to check for the existence of config?

[Kalyan] Good catch.. Fixed..

> The DownloaderStatus check has the same issue as the VerifyImageStatus check hence I'd suggest to make those checks be identical using functions with a consistent naming scheme.

[Kalyan] Good Catch.. Fixed..

eriknordmark commented on pkg/pillar/cmd/zedmanager/updatestatus.go:
```diff
@@ -371,10 +371,85 @@ func doUpdate(ctx *zedmanagerContext, uuidStr string,
+			safename, ssPtr.ImageSha256)
+		return nil, false
+	}
+	if vs.Expired {
```
Close but not quite right.
Expired means that the verifier is asking zedmanager whether the verifier can delete the image.
The zedmantager then checks the config refcount and if zero tells the verifier it can go ahead and delete the image by deleting VerifyImageConfig.
Thus here you should check if the VerifyImageConfig has been deleted.

eriknordmark replied:
Based on further discussing there is a race condition since a VerifyImageConfig can in theory be created after it has been deleted and before the verifier has run and deleted the VerifyImageStatus. Hence the code in the PR is fine.

eriknordmark commented on pkg/pillar/types/verifiertypes.go:
```diff
@@ -85,6 +85,45 @@ func (status VerifyImageStatus) VerifyFilename(fileName string) bool {
+
+// ImageFilenames - Returns pendingFilename, verifierFilename, verifiedFilename
+// for the image
+func (status VerifyImageStatus) ImageFilenames() (string, string, string) {
```
Either both functions should have "Download" in their names, or neither should have it.
They return the directory vs filename for the same thing hence that difference in naming is confusing.

kalyan-nidumolu replied:
Ah... Good catch.. Fixed it..
